s3fs/core.py

```diff
@@ -355,11 +368,12 @@ def rm(self, path, recursive=False):
                 raise IOError('Delete key failed', (bucket, key))
             self._ls(path, refresh=True)
         else:
-            if recursive or not self.s3.list_objects(Bucket=bucket).get('Contents'):
+            if not self.s3.list_objects(Bucket=bucket).get('Contents'):
                 try:
                     self.s3.delete_bucket(Bucket=bucket)
                 except ClientError:
                     raise IOError('Delete bucket failed', bucket)
+                self.dirs.pop(bucket, None)
```

Comment:
If you delete a bucket, the entry will still be in the `dirs` cache (although it should be empty), so `ls(mybucket)` would not produce the not-found error that it should.